### PR TITLE
frontend-feat: #5512 Added metro latest changes

### PR DESCRIPTION
**Pull request by spravinkumar9952**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [x] Enhancement
- [x] Refactoring
- [ ] Dependency updates

## Description

1. Payment SDK pooling in metro
2. Ticket status api call in mytickets screen
4. UI fixes
5. Back press issue fixes

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Frontend/ui-customer/src/Flow.purs [hunk 1]
```diff
@@ -3388,20 +3385,33 @@ metroTicketDetailsFlow = do
...
-      metroTicketDetailsFlow
+    GO_TO_METRO_TICKET_DETAILS_FLOW  bookingId -> do
+      (GetMetroBookingStatusResp resp) <- Remote.getMetroStatusBT bookingId
+      let (MetroTicketBookingStatus metroTicketBookingStatus) = resp
+      if(metroTicketBookingStatus.status == "CONFIRMED") then do
+        modifyScreenState $ MetroTicketDetailsScreenStateType (\metroTicketDetailsState -> metroTicketDetailsTransformer resp metroTicketDetailsState)
+        modifyScreenState $ MetroTicketDetailsScreenStateType (\slsState -> slsState{props{previousScreenStage = ST.MetroMyTicketsStage}})
+        metroTicketDetailsFlow
+      else if (metroTicketBookingStatus.status == "PAYMENT_PENDING" || metroTicketBookingStatus.status == "FAILED") then do
+        modifyScreenState $ MetroTicketStatusScreenStateType (\metroTicketStatusScreen -> metroTicketStatusTransformer resp "" metroTicketStatusScreen)
```
Comment (aryan-13) [inline Frontend/ui-customer/src/Flow.purs:3404]: No need to pass empty string to `metroTicketStatusTransformer`, If it is not being used.

File: Frontend/ui-customer/src/Flow.purs [hunk 2]
```diff
@@ -3105,18 +3106,19 @@ metroTicketBookingFlow = do
...
-          (confirmMetroQuoteResp) <- Remote.confirmMetroQuoteBT state.data.quoteId
-          let (MetroTicketBookingStatus metroBookingStatus) = confirmMetroQuoteResp
-          modifyScreenState $ MetroTicketBookingScreenStateType (\state -> state{data{ bookingId = ( (metroBookingStatus.bookingId))}})
-          void $ lift $ lift $ delay $ Milliseconds 3000.0
-          metroTicketPaymentFlow metroBookingStatus.bookingId
-          else do
-            metroTicketBookingFlow  
+      else if state.props.currentStage == ConfirmMetroQuote then do
+        (confirmMetroQuoteResp) <- Remote.confirmMetroQuoteBT state.data.quoteId
+        let (MetroTicketBookingStatus metroBookingStatus) = confirmMetroQuoteResp
```
Comment (aryan-13) [inline Frontend/ui-customer/src/Flow.purs:3111]: update this to `(MetroTicketBookingStatus confirmMetroQuoteResp) <- Remote.confirmMetroQuoteBT state.data.quoteId`

Then `Line 3111` is not required

**Review by aryan-13 — CHANGES_REQUESTED** [review]

File: Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroTicketDetails/View.purs [hunk 3]
```diff
@@ -349,7 +354,7 @@ qrCodeView push state =
       , imageView [
           width $ V 32
         , height $ V 32
-        , visibility $ boolToVisibility $ state.data.noOfTickets > 0
+        , visibility $  if state.data.noOfTickets > 1 then VISIBLE else INVISIBLE
```
[the remark this answers is not in the document]
Comment (spravinkumar9952) [reply Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroTicketDetails/View.purs:357]: here i am using INVISIBLE not GONE. so it won't work here

[the remark this answers is not in the document]
Comment (aryan-13) [reply Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroTicketDetails/View.purs:357]: Yes bro, we have `boolToInvisibility` for `INVISIBLE` and `boolToVisibility` for `GONE`. Please check once

File: Frontend/ui-customer/src/Flow.purs [hunk 4]
```diff
@@ -3411,7 +3421,8 @@ metroTicketStatusFlow = do
   case flow of 
     GO_TO_METRO_TICKET_DETAILS state resp -> do
       modifyScreenState $ MetroTicketDetailsScreenStateType (\_ -> MetroTicketDetailsScreenData.initData)
-      modifyScreenState $ MetroTicketDetailsScreenStateType (\metroTicketDetailsState -> metroTicketDetailsTransformer resp metroTicketDetailsState)
+      modifyScreenState $ MetroTicketDetailsScreenStateType (\metroTicketDetailsState -> metroTicketDetailsTransformer resp metroTicketDetailsState) 
+      modifyScreenState $ MetroTicketDetailsScreenStateType (\slsState -> slsState{props{previousScreenStage = ST.MetroTicketStatusStage}})
```
[the remark this answers is not in the document]
Comment (spravinkumar9952) [reply Frontend/ui-customer/src/Flow.purs:3425]: since we don't have the state for the transformer, we can only reduce to 2.

File: Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroMyTickets/Transformer.purs [hunk 5]
```diff
@@ -45,16 +57,18 @@ ticketItemTransformer (MetroTicketBookingStatus bookingItem) =
   let 
     sourceStationEnum = bookingItem.stations !! 0
     destinationStationEnum = bookingItem.stations !! ((length bookingItem.stations)- 1)
-    ticketsValidTill =  case bookingItem.tickets !! 0 of
-                              Just (FRFSTicketAPI ticket) -> (convertUTCtoISC ticket.validTill "hh:mm A") <> ", " <> (convertUTCtoISC ticket.validTill "Do MMM YYYY") 
-                              Nothing -> ""
+    utcValidTill =  maybe "" (\(FRFSTicketAPI ticket) -> ticket.validTill) (bookingItem.tickets !! 0)
```
Comment (aryan-13) [inline Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroMyTickets/Transformer.purs:60]: We can use `head` here

File: Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroTicketBooking/View.purs [hunk 6]
```diff
@@ -60,14 +66,18 @@ screen initialState =
...
+          ShowMetroBookingTimeError _ -> pure unit
+          _ -> void $ pure $ setValueToLocalStore METRO_PAYMENT_SDK_POLLING "false"
         eval action state
   }
   where
     getQuotes push = do
       let withinTimeRange = JB.withinTimeRange "04:30:00" "22:30:00" $ EHC.convertUTCtoISC (EHC.getCurrentUTC "") "HH:mm:ss"
       push $ ShowMetroBookingTimeError withinTimeRange
       void $ launchAff $ EHC.flowRunner defaultGlobalState $ getQuotesPolling initialState.data.searchId 5 3000.0 initialState push GetMetroQuotesAction
+      void $ launchAff $ EHC.flowRunner defaultGlobalState $ runExceptT $ runBackT $ getSDKPolling initialState.data.bookingId 3000.0 initialState push GetSDKPoolingAC
```
Comment (kavya-shree-s) [inline Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroTicketBooking/View.purs:80]: GetSDKPoolingAC rename to GetSDKPollingAC

File: Frontend/ui-customer/src/Flow.purs [hunk 7]
```diff
@@ -3388,20 +3384,33 @@ metroTicketDetailsFlow = do
...
   case flow of 
-    GO_TO_METRO_TICKET_DETAILS_FLOW bookingStatusResp -> do
-      modifyScreenState $ MetroTicketDetailsScreenStateType (\metroTicketDetailsState -> metroTicketDetailsTransformer bookingStatusResp metroTicketDetailsState)
-      modifyScreenState $ MetroTicketDetailsScreenStateType (\slsState -> slsState{props{previousScreenStage = ST.MetroTicketStatusStage}})
-      metroTicketDetailsFlow
+    GO_TO_METRO_TICKET_DETAILS_FLOW  bookingId -> do
+      (GetMetroBookingStatusResp resp) <- Remote.getMetroStatusBT bookingId
+      let (MetroTicketBookingStatus metroTicketBookingStatus) = resp
+      if(metroTicketBookingStatus.status == "CONFIRMED") then do
+        modifyScreenState $ MetroTicketDetailsScreenStateType (\metroTicketDetailsState -> metroTicketDetailsTransformer resp metroTicketDetailsState)
```
Comment (aryan-13) [inline Frontend/ui-customer/src/Flow.purs:3399]: We can combine these 2 modifyScreenState calls as well here and in the `else` case well

File: Frontend/ui-customer/src/Flow.purs [hunk 8]
```diff
@@ -3126,51 +3127,46 @@ metroTicketBookingFlow = do
...
+  let (PaymentPagePayload sdk_payload) = orderResp.sdk_payload
+      (PayPayload innerpayload) = sdk_payload.payload
+      finalPayload = PayPayload $ innerpayload{ language = Just (getPaymentPageLangKey (getLanguageLocale languageKey)) }
+      sdkPayload = PaymentPagePayload $ sdk_payload{payload = finalPayload}
+      shortOrderID = orderResp.order_id
+  lift $ lift $ doAff $ makeAff \cb -> runEffectFn1 checkPPInitiateStatus (cb <<< Right) $> nonCanceler
+  _ <- paymentPageUI sdkPayload
+  void $ lift $ lift $ loaderText (getString STR.LOADING) (getString STR.PLEASE_WAIT_WHILE_IN_PROGRESS)
+  void $ lift $ lift $ toggleLoader true
+  void $ lift $ lift $ delay $ Milliseconds 3000.0
```
Comment (kavya-shree-s) [inline Frontend/ui-customer/src/Flow.purs:3153]: @spravinkumar9952 This delay can be removed

**Review by aryan-13 — CHANGES_REQUESTED** [review]

File: Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroTicketBooking/View.purs [hunk 9]
```diff
@@ -60,14 +66,18 @@ screen initialState =
   , globalEvents : [getQuotes]
   , eval : \action state -> do
         let _ = spy "MetroTicketBookingScreenState action " action
-        let _ = spy "MetroTicketBookingScreenState state " state
+        let _ = spy "MetroTicketBookingScreenState state " state 
+        case action of
+          ShowMetroBookingTimeError _ -> pure unit
```
Comment (kavya-shree-s) [inline Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroTicketBooking/View.purs:71]: @spravinkumar9952 Why was this required `ShowMetroBookingTimeError` ?

File: Frontend/ui-customer/src/Screens/Types.purs [hunk 10]
```diff
@@ -2305,4 +2309,7 @@ type MetroTicketStatusScreenData = {
 type MetroTicketStatusScreenProps = {
   showShimmer :: Boolean
 , paymentStatus :: PP.PaymentStatus
+, entryPoint :: MetroTicketStatusScreenEntry
 }
+
+data MetroTicketStatusScreenEntry = HomescreenToMetroTicketStaus | MyMetroTicketsToMetroTicketStatus
```
Comment (kavya-shree-s) [inline Frontend/ui-customer/src/Screens/Types.purs:2315]: If typo HomescreenToMetroTicketStatus

**Review by aryan-13 — DISMISSED** [review]

File: Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroMyTickets/Transformer.purs [hunk 11]
```diff
@@ -22,20 +22,32 @@ import Services.API
...
       , pastTickets = pastTickets'
       }
     }
+  where 
+    activeTicketEvaluator :: String -> Array FRFSTicketAPI  -> Boolean
+    activeTicketEvaluator status tickets = 
+      let validTill = maybe "" (\(FRFSTicketAPI ticket) -> ticket.validTill) (head tickets)
+      in (any (_ == status) ["CONFIRMED", "PAYMENT_PENDING"]) && (not $ isTicketExpired validTill)
+
+    pastTicketEvaluator :: String -> Array FRFSTicketAPI -> Boolean
```
Comment (kavya-shree-s) [inline Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroMyTickets/Transformer.purs:46]: @spravinkumar9952 sort tickets based on time is not handled

**Review by aryan-13 — APPROVED** [review]

**Review by kavya-shree-s — CHANGES_REQUESTED** [review]

File: Frontend/ui-customer/src/Flow.purs [hunk 12]
```diff
@@ -1579,6 +1579,7 @@ homeScreenFlow = do
     GO_TO_MY_METRO_TICKETS -> do
       (GetMetroBookingListResp resp)<- Remote.getMetroBookingStatusListBT
```
Comment (kavya-shree-s) [inline Frontend/ui-customer/src/Flow.purs:1580]: This api call can be removed from flow as it is moved to global event

File: Frontend/ui-customer/src/Flow.purs [hunk 13]
```diff
@@ -3105,18 +3106,18 @@ metroTicketBookingFlow = do
...
+      else if state.props.currentStage == ConfirmMetroQuote then do
+        (MetroTicketBookingStatus metroBookingStatus) <- Remote.confirmMetroQuoteBT state.data.quoteId
+        modifyScreenState $ MetroTicketBookingScreenStateType (\state -> 
+          state{data{ bookingId = metroBookingStatus.bookingId}
+              , props {currentStage = ST.PaymentSDKPooling }})
+        setValueToLocalStore METRO_PAYMENT_SDK_POLLING "true"
+      else pure unit 
+      metroTicketBookingFlow  
     GO_TO_MY_METRO_TICKET_SCREEN -> do
       (GetMetroBookingListResp resp)<- Remote.getMetroBookingStatusListBT
```
Comment (kavya-shree-s) [inline Frontend/ui-customer/src/Flow.purs:3118]: Can we  remove this api call  from flow as it is moved to global event

File: Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroTicketBooking/View.purs (hunk 9 above)
Comment (spravinkumar9952) [reply Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroTicketBooking/View.purs:71]: yes. otherwise pooling will stop

File: Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroMyTickets/Transformer.purs (hunk 11 above)
Comment (spravinkumar9952) [reply Frontend/ui-customer/src/Screens/TicketBookingFlow/MetroMyTickets/Transformer.purs:46]: coming in descending order from backend. so reversed the array

**Review by kavya-shree-s — APPROVED** [review]
